### PR TITLE
Tune razor and futility margins

### PR DESCRIPTION
**Pull request by lantonov**
Tuning was done with Bayesian optimisation with the following parameters:
Acquisition function: Expected Improvement
alpha: 0.05
xi: 1e-4
TC: 60+0.6
Number of iterations: 100
Initial points: 5
Batch size: 20 games

STC
http://tests.stockfishchess.org/tests/view/5dee291e3cff9a249bb9e470
LLR: 2.97 (-2.94,2.94) [-1.50,4.50]
Total: 19586 W: 4382 L: 4214 D: 10990

LTC
http://tests.stockfishchess.org/tests/view/5dee4e273cff9a249bb9e473
LLR: 2.95 (-2.94,2.94) [0.00,3.50]
Total: 38840 W: 6315 L: 6036 D: 26489

Bench 5767220

### CODE REVIEW

**Comment by snicolet** [comment]
@lantonov That's qui impressive, congrats! 

Merged via 443787b0d1dceb6186e217ea2b2224e76806ea15, but I will keep the pull request open for a while because I think it would be interesting if you could tell us a little bit more about the Bayesian optimisation method you used — I am not sure everybody in the team is up-to-date with the theory and tools you used.

**Comment by lantonov** [comment]
For this optimisation I used a combination of techniques, one part of which is Bayesian optimisation and the other part is modeling of ordinal category variables. I am not sure if we can do without the modeling part, maybe just with some combination of acquisition functions / kernels but here we are, I am still trying and nothing is set in stone.
Bayesian optimisation (BO) is a technique for optimising black-box functions. By black-box it is meant that nothing is known or assumed about these functions: they may be concave or not concave, continuous or not continuous, and so on. These functions may have one (easy!) or more arguments (parameters) but usually not more than 20 (> 20x20 matrices or kernels become too big and unwieldy). Evaluation of the function is restricted to sampling at some point in the d-dimensional space (d is number of parameters) and getting a possibly noisy response. If the function is cheap to evaluate we could sample at many points e.g. via grid search, random search or numeric gradient estimation. However, if function evaluation is expensive e.g. taking a sample of 20-30-40 games, then it is important to minimize the number of samples drawn from the black box function. BO attempts to find the global optimimum in a minimum number of steps. BO incorporates prior belief about the function and updates the prior with samples drawn from it to get a posterior that better approximates the function. The model used for approximating the objective function is called surrogate model. BO also uses an acquisition function that directs sampling to areas where an improvement over the current best observation is likely.
A popular surrogate model for BO are Gaussian processes (GP). GP is a random process where any point x is assigned a random variable f(x) and where the joint distribution of a finite number of these variables p(f(x1),…,f(xN)) is itself Gaussian:
image with 
image and 
image
_m_ is the mean function and it is common to use _m_(x)=0 as GPs are flexible enough to model the mean arbitrarily well. κ is a positive definite kernel function or covariance function. Thus, a Gaussian process is a distribution over functions whose shape (smoothness, …) is defined by **K**, the kernel. If points x_i and x_j are considered to be similar by the kernel the function values at these points, f(x_i) and f(x_j), can be expected to be similar too. A GP prior _p_(_f_|**X**) can be converted into a GP posterior _p_(_f_|**X**,**y**) after having observed some data **y**. The posterior can then be used to make predictions **f**∗ given new input **X**∗:
image
This is the posterior predictive distribution which is also a Gaussian with mean μ∗ and **Σ**∗ (the covariance matrix). By definition of the GP, the joint distribution of observed data **y** and predictions **f**∗ is
image
With N training data and N∗ new input data, **K**_y=κ(**X**,**X**)+σ^2_y**I**=**K**+σ^2_y**I** is N×N, **K**∗=κ(**X**,**X**∗) is N×N∗ and **K**∗∗=κ(**X**∗,**X**∗) is N∗×N∗. σ^2_y is the noise term in the diagonal of **K**_y. It is set to zero if training targets are noise-free and to a value greater than zero if observations are noisy. The mean is set to 0 for notational simplicity. The sufficient statistics of the posterior predictive distribution, μ∗ and **Σ**∗, can be computed with
image
The GP posterior is cheap to evaluate and is used to propose points in the search space where sampling is likely to yield an improvement. 
The kernels **K** typically have the property that points closer to the input space are more strongly correlated. Additionally, kernels are required to be positive semi-definite functions. A simple and commonly used kernel is the power exponential or Gaussian kernel,
image where 
image and alpha[0:d] are parameters of the kernel. Below are shown 3 random functions drawn from a GP with a Gaussian kernel
image
Each plot corresponds to a different value for the parameter alpha, with alpha decreasing from left to right. Varying this parameter creates different beliefs about how quickly f(x) changes with x.
I have used primarily another kernel which is believed to be more noise-resistant, the so-called Matern kernel: 
image where K_nu is the modified Bessel function, and we have a parameter nu in addition to the parameters alpha[0:d]. The constant mean function eta of a Matern kernel depends on alpha[0:d], nu, and mu, and the last one (mu) is usually chosen to be 0. eta is determined from the data by one of 3 ways: (i) maximum likelihood estimate (MLE), (ii) maximum a posteriori estimate (MAP), and (iii) fully Bayesian approach whereby we compute the posterior distribution on f(x) marginalizing over all possible values of the hyperparameters approximating the respective intractable integral by a sum obtained through sampling.
The BO procedure is as follows. For t=1,2,… repeat:
1. Find the next sampling point x_t by optimizing the acquisition function over the GP: x_t=argmax_x u(x|D_1:t−1)
2. Obtain a possibly noisy sample y_t=f(x_t)+ϵ_t from the objective function f.
3. Add the sample to previous samples D_1:t={D_1:t−1,(x_t,y_t)} and update the GP.
There are several acquisition functions and 3 of them are most widely used: Expected Improvement (EI), Knowledge Gradient (KG), and Predicted Entropy Search (PES). I have used exclusively EI and at latter stages I intend to use also the other AFs or a combination of them.
EI is defined as 
image where f(x+) is the value of the best sample so far and x+ is the location of that sample. The expected improvement can be evaluated analytically under the GP model: 
image where 
image where μ(x) and σ(x) are the mean and the standard deviation of the GP posterior predictive at x, respectively. Φ and ϕ are the CDF and PDF of the standard normal distribution, respectively. The first summation term in the above formula for EI is the exploitation term and second summation term is the exploration term. Parameter ξ determines the amount of exploration during optimization and higher ξ values lead to more exploration. In other words, with increasing ξ values, the importance of improvements predicted by the GP posterior mean μ(x) decreases relative to the importance of potential improvements in regions of high prediction uncertainty, represented by large σ(x) values. A default value for ξ is 0.01. I found that smaller xi (e.g. 1e-4) gives somewhat better results given the uncertainty of the maximal estimate generated by sampling with different opening sets.
This in brief is the theory of BO. I will try to find some more time for the implementation details, and, eventually, for the modeling approach(es) that I have used.

**Comment by joergoster** [comment]
@lantonov What I find most amazing is that this worked with only 20 games per iteration!

When I did my experiments with BO with [this](https://github.com/thomasahle/fastchess/tree/master/tools) implementation by @thomasahle, after many unsuccessful attempts, it happened to make sense that the number of games you need to play per iteration is highly dependent on the expected elo gain. So the higher the expected gain, the lower the number of games you have to play. However, I do consider something like 100 or 200 games as an absolute minimum. There is just too much noise involved in chess games!

It also helped to return the mean of the best evaluated points instead of simply the best one. This was inspired by

> Return a solution: either the point evaluated with the largest f (x), or the point with the largest posterior mean.

which you can find on top of page 3 in this paper: https://arxiv.org/abs/1807.02811#

**Comment by vondele** [comment]
yes, I share the experience @joergoster is talking about... in the past, I tried this with 1000s of games per point, with no luck. Hopefully, this positive experience is repeatable!?

**Comment by lantonov** [comment]
@joergoster Thanks for pointing me to sources I was not aware of. 
In my tests I have tried many different numbers of games per iteration, from 2 to 1000. I found that 20 to 40 games per iteraton leverage best the balance number of games per iteration : total number of iterations. To a large part this is due to the fact that BO allows adjustment exploration vs exploitation, and you can adjust it so that points with big f(x) receive more attention than the others, i.e. these points or points close to them are repeated more often. Then the GP averages the multiple f(x) at these points in such a way that it takes into account also the f(x) at other points. Nevertheless this requires a lot of patience. For example, finding eta for the Matern kernel took me about 2 months, and this eta is applicable only for the specific optimisation (in my case, the 4 safechecks). This patch (2 parameters) required a different eta which took me an additional month. This is why I resorted to post-BO modeling of ordinal categorical variables in which I have more experience. Another thing is that my scoring of the sample is different. Rather than using a scalar for the score, I use a vector (a list) which can also be looked at as a pentanomial distribution . For example, with a sample of 20 games, the score can look like [0,2,4,3,1] which contains much more information than a simple scalar. This 5-category list can be used to calculate marginal probabilities through forming ridits and various logits or probits (cumulative, adjacent, global, proportional, etc.). This allows you to model all the samples as whole, and not only from iteration to iteration. Still, I keep on tinkering trying to do BO in such a way that it is possible to avoid the modeling for which a specialised statistical software is necessary (R, SAS).

**Comment by adentong** [comment]
I gotta say, this is amazing stuff! Though I don't understand any of this stuff, it's still really cool to see people engaging in theoretical discussions (in a PR for a chess engine no less).

**Comment by MichaelB7** [comment]
Same here - fascinating stuff.  A little over my head for sure.

**Comment by lantonov** [comment]
The question of how many games per iteration is really important one. We can do the following experiment (I did it some time ago but have lost the results). Say, we have a budget of 1000 games and we must distribute them to iterations. We can make 50 iterations 20 games each, 20 iterations 50 games each, and 10 iterations 100 games each. We can put on both sides one and the same SF (exact match). In this case we know for sure that in the long run (asymptotically), with infinite number of games, the result will be exactly 50:50, i.e. we have a 100% probability for an even result. This experiment has 2 aspects. The first one is descriptive statistics. We can do a comparison of 2 measures: LOS by normal approximation, and LOS by non-parametric statistics (e.g. through ridits). This will show how well CLT (Central Limit Theorem) performs with the different sample sizes. As a bonus we will also get reliable measures of the accuracy (deviation from 50%) and precision (the inverse of the standard deviation). We can play further with different experimental settings to look for sources of variation. Thus, an ANOVA can tell us the difference between playing from one position (say, the startting position) and playing from randomly chosen positions. This first aspect is certain to favour the larger sample sizes: you will get better accuracy and precision with them.
The second aspect is more complicated and involves 2 different engines: the first one with the default settings, and the second one with equally spaced settings within the bounds of some parameter for each iteration.  In addition to accuracy and precision, a very important criterion here is coverage: you can get much better coverage with 50 points than with 10 points, albeit with lower accuracy and precision for the 50 points case. The results from this semi-controlled experiment can be put in the various matrices of the Bayesian optimisation and so we can assess how this method (BO) performs with different sample sizes and choose the optimal one.

**Comment by gonzalezjo** [comment]
Would you be able to share the code you used to tune?

**Comment by lantonov** [comment]
The more important software I have made for optimisation is in [my github repository](https://github.com/lantonov/Optimisation). For a tc BO, bayesopt_cli.py and cutechess_batch.py are enough.

**Comment by joergoster** [comment]
@lantonov Thank you for this very interesting, additional information.

Scikit-optimize (skopt) only supports scalars in its implementation as evaluations, afaik. Unfortunately, the whole project is no more under active development. When I have time, I will definitely take a closer look at https://github.com/fmfn/BayesianOptimization.

**Comment by lantonov** [comment]
For the Gaussian processes there is a [free book](http://www.gaussianprocess.org/gpml/) published by MIT Press which is very detailed.

**Comment by snicolet** [comment]
Do you know this tool, which is a library of gradient-free optimization methods? 
Nevergrad: An open source tool for derivative-free optimization
https://engineering.fb.com/ai-research/nevergrad/

**Comment by joergoster** [comment]
@snicolet Thanks! Nevergrad seems to be actively maintained, and it also provides a ask and tell interface. Definitely worth a look!

**Comment by lantonov** [comment]
@snicolet Thank you! Have not seen it before. Just a passing glance alerted me to things deserving a closer attention. Useful resource!

**Comment by gonzalezjo** [comment]
I see that on [the webpage for your repository](https://lantonov.github.io/Optimisation/) you mention that "the algorithm in this program ... is shown to have better convergence than [SPSA]."


How much of an engineering effort do you believe it would be to replace (or add as an alternative to) SPSA with this algorithm on fishtest? The small batch-sizes seem a bit concerning for something like fishtest, but I'm not sure if it's as big of an issue as I think it is.

**Comment by lantonov** [comment]
The sentence that you quoted concerns another algorithm -- ROCK*, which uses a gradient (like SPSA). I tried ROCK* 5 years ago but it turned out that a continuous and convex objective function is important for it as for most of the gradient-based algorithms. It is hard or impossible to transform chess results to such a function so ROCK* did not work -- the inverted covariance matrix turned either zero or infinity after 100 iterations or so.
As concerns Bayesian optimisation, it is promising because it does not have any requirements or assumptions about the objective function, and in addition, it tolerates noise. For the implementation of BO, it is important to assess the noise of our game results and tune the BO hyperparameters. It is not easy and requires a lot of trial and error. It is hard to give any definite deadline for it. BO should become easy to do and reproducible enough to become an alternative to SPSA.

**Comment by lantonov** [comment]
Very relevant paper just out: [Ordinal Bayesian Optimisation](https://arxiv.org/abs/1912.02493). Hope it helps obviate the necessity for ordinal modelling.

**Comment by miguel-l** [comment]
Interesting work! I dream that this leads to eventually make more optimization methods available via fishtest.

**Comment by snicolet** [comment]
See https://groups.google.com/forum/?fromgroups=#!topic/fishcooking/b3uhBBJcJG4

We can continue the discussion in the above fishcooking forum thread, so it seems I can close this PR now.
